src/Storages/HDFS/StorageHDFS.cpp

```diff
@@ -72,9 +73,17 @@ namespace
 
         HDFSFileInfo ls;
         ls.file_info = hdfsListDirectory(fs.get(), prefix_without_globs.data(), &ls.length);
+        if (ls.file_info == nullptr && errno != ENOENT) // NOLINT
+        {
+            // ignore file not found exception, keep throw other exception, libhdfs3 doesn't have function to get exception type, so use errno.
+            throw Exception(
+                ErrorCodes::ACCESS_DENIED, "Cannot list directory {}: {}", prefix_without_globs, String(hdfsGetLastError()));
+        }
         Strings result;
         for (int i = 0; i < ls.length; ++i)
         {
+            if (ls.file_info == nullptr)
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "file_info shouldn't be null");
```

Comment:
It is better to move this check outside of loop and under `if (!ls.file_info && ls.length > 0) throw`

Reply:
done